### ISSUE TITLE
Allow access to V8Locker.thread and improve "invalid thread access" messages

### ISSUE THREAD

**Issue by drywolf** [eclipsesource/J2V8#228]
Hi,

I am trying to use J2V8 from within a server platform (SonarQube) which makes much use of multi-threading. Therefore it is quite a tricky thing to not violate the single threading access when interacting with J2V8.

To make it easier to reason about what is going on when using the V8Locker class to acquire and release access of a thread to a V8 runtime, I would propose:

- to allow read access to the V8Locker.thread field that stores the currently assigned thread
- improve the message of [acquire()](https://github.com/eclipsesource/J2V8/blob/c97922e95c3ba7d6e544f81bcf6bf14b788e03a0/src/main/java/com/eclipsesource/v8/V8Locker.java#L34) to tell the developer that acquisition of the locker failed and which threads are involved (V8Locker thread / thread requesting lock acquisition)
- improve the message of [checkThread()](https://github.com/eclipsesource/J2V8/blob/c97922e95c3ba7d6e544f81bcf6bf14b788e03a0/src/main/java/com/eclipsesource/v8/V8Locker.java#L56) to tell the developer that the locker ownership-check failed and which threads are involved (V8Locker thread / thread checking lock ownership)
- provide non-throwing versions of the above two methods for use in regular program-logic rather than the exception case

Please let me know what you think of those, it would really help me out to get things working much more effortlessly and quickly for what I'm trying.

I would provide a pull request once you can confirm that my proposed additions are a good idea.

Regards

**Comment by Mizumi** [comment]
A [pull request that's currently pending a merge](https://github.com/eclipsesource/J2V8/pull/201) partially addresses this with the introduction of a ConcurrentV8 class (which you can see [here](https://github.com/Mizumi/J2V8/blob/34c2bc7986d651a7419e37b2fc4dc7b174338605/src/main/java/com/eclipsesource/v8/ConcurrentV8.java)). 

Maybe the ConcurrentV8 functionality could be amended by or merged with your suggestions, or even better it may meet your needs entirely. ;D

**Comment by drywolf** [comment]
Thanks for the reply.

I had a quick look at the PR and saw that it is not only about threading but also about JAVA<->JS interop with classes ... which is exactly what I'm trying to do currently :smile:

I also saw in the PR discussion that [J2V8-classes](https://github.com/jsio-private/J2V8-classes) was mentioned which is what I am currently using and fixing/extending as I go along.

@irbull @Mizumi @yofreke @drywolf
I think it would be great to have a joined discussion of what the supported features / usecases should be and how J2V8 could best support them, and also to consolidate the work that is being done.

For example I integrated some missing / incomplete functionality in J2V8-classes that should allow to instantiate classes via Java reflection that were extened in a JS script from a base java class/interface.

Let me know what you guys are thinking, I would be happy to collaborate on this.

**Comment by Mizumi** [comment]
Glad it helped you! Yeah, a lot of people have been trying to do similar things across the board. The only thing my PR doesn't address is extending Java classes in JS. 

However, it does support the implementation of Java functional interfaces (lambdas) with zero extra effort on either end of the API via the use of proxy classes. While this is definitely my own opinion showing through (and not in any way a fact), I'm of the belief that's it's better to avoid extending entire classes from Ecma-like script run times since it simplifies the architecture of the system, as well as avoids egregious hacks in the JS side. 

That said, I think it would be trivial to add the capability for entire class extension to meet everyone's needs. I'd be delighted to have a discussion about these features so we can get them incorporated directly into V8.

**Comment by irbull** [comment]
Thanks everyone.

I have added some of these to master. In particular:
1. I have provided access to the underlying `thread` field with dc823239669be6650f1351a81c01c9e979b001a0
2. I have improved the thread access error message with ffaced355c44e9262bd6be983036e98b0397ef8a
3. I have implemented a `tryAcquire` that does not throw an exception, but rather returns true or false depending on if the lock could be acquired with 53981c165b6532084c4158fd1e6ff8f785d3328f

We already have `hasLock` so I don't think we need a non-throwing version of `checkThread`. I'm going to close this issue. @drywolf let me know if these work for you, or if you think we need anything else on the `V8Locker`.
